api/layer/multipart_upload.go

```diff
@@ -201,35 +206,120 @@ func (n *layer) uploadPart(ctx context.Context, multipartInfo *data.MultipartInf
 		}
-		prm.Attributes = append(prm.Attributes, [2]string{AttributeDecryptedSize, strconv.FormatInt(p.Size, 10)})
-		prm.Payload = r
+		attributes = append(attributes, [2]string{AttributeDecryptedSize, strconv.FormatInt(p.Size, 10)})
+		payloadReader = r
 		p.Size = int64(encSize)
 	}
 
-	prm.Attributes[0][0], prm.Attributes[0][1] = UploadIDAttributeName, p.Info.UploadID
-	prm.Attributes[1][0], prm.Attributes[1][1] = UploadPartNumberAttributeName, strconv.Itoa(p.PartNumber)
+	var (
+		splitPreviousID      oid.ID
+		isSetSplitPreviousID bool
+		multipartHash        = sha256.New()
+		tzHash               hash.Hash
+	)
 
-	id, hash, err := n.objectPutAndHash(ctx, prm, bktInfo)
+	if n.neoFS.IsHomomorphicEnabled() {
+		tzHash = tz.New()
+	}
+
+	lastPart, err := n.treeService.GetLastPart(ctx, bktInfo, multipartInfo.ID)
 	if err != nil {
-		return nil, err
+		// if ErrPartListIsEmpty, there is the first part of multipart.
+		if !errors.Is(err, ErrPartListIsEmpty) {
+			return nil, fmt.Errorf("getLastPart: %w", err)
+		}
+	} else {
+		// try to restore hash state from the last part.
+		if binaryUnmarshaler, ok := multipartHash.(encoding.BinaryUnmarshaler); ok {
+			if err = binaryUnmarshaler.UnmarshalBinary(lastPart.MultipartHash); err != nil {
+				return nil, fmt.Errorf("unmarshalBinary: %w", err)
+			}
+		}
+
+		if n.neoFS.IsHomomorphicEnabled() {
+			if binaryUnmarshaler, ok := tzHash.(encoding.BinaryUnmarshaler); ok {
+				if err = binaryUnmarshaler.UnmarshalBinary(lastPart.HomoHash); err != nil {
+					return nil, fmt.Errorf("unmarshalBinary: %w", err)
+				}
+			}
+		}
+
+		isSetSplitPreviousID = true
+		splitPreviousID = lastPart.OID
 	}
 
+	var (
+		id           oid.ID
+		elements     []string
+		creationTime = TimeNow(ctx)
+		// User may upload part large maxObjectSize in NeoFS. From users point of view it is a single object.
+		// We have to calculate the hash from this object separately.
+		currentPartHash = sha256.New()
+	)
+
+	multipartHashes := []hash.Hash{multipartHash, currentPartHash}
+	if tzHash != nil {
+		multipartHashes = append(multipartHashes, tzHash)
+	}
+
+	prm := PrmObjectCreate{
+		Container:       bktInfo.CID,
+		Creator:         bktInfo.Owner,
+		Attributes:      attributes,
+		CreationTime:    creationTime,
+		CopiesNumber:    multipartInfo.CopiesNumber,
+		SplitID:         multipartInfo.SplitID,
+		MultipartHashes: multipartHashes,
+	}
+
+	dt := n.buffers.Get()
+	chunk := dt.(*[]byte)
+
+	// slice part manually. Simultaneously considering the part is a single object for user.
+	for {
+		if isSetSplitPreviousID {
+			prm.SplitPreviousID = &splitPreviousID
+		}
+
+		nBts, readErr := payloadReader.Read(*chunk)
+		if nBts > 0 {
+			prm.Payload = bytes.NewReader((*chunk)[:nBts])
+			prm.PayloadSize = uint64(nBts)
+
+			id, _, err = n.objectPutAndHash(ctx, prm, bktInfo)
+			if err != nil {
+				return nil, err
+			}
+
+			isSetSplitPreviousID = true
+			splitPreviousID = id
+			elements = append(elements, id.String())
```

Comment:
according to the usage of this var it seems to me that storing just raw ID is more efficient

Reply:
Updated